state/fork/traverse.go

```diff
@@ -0,0 +1,117 @@
+	lowestHeightToVisit, sanityChecker, err := terminal.Translate2Height(headers)
+	if err != nil {
+		return fmt.Errorf("error evaluating terminal: %w", err)
+	}
+
+	startBlock, err := headers.ByBlockID(forkHead)
+	if err != nil {
+		return fmt.Errorf("could not fork head with if  %x: %w", forkHead, err)
+	}
+	if startBlock.Height < lowestHeightToVisit {
+		return nil
+	}
+
+	return unsafeTraverse(headers, startBlock, visitor, lowestHeightToVisit, sanityChecker)
+}
+
+// unsafeTraverse implements the fork traversal in the order of decreasing height.
+// It is unsafe because:
+// * always calls the `visitor` on the `block`, _before_ it checks whether to stop
+// * if `block` has a lower height than lowestHeightToVisit, the traversal keeps
+//   going until we hit the root block
+func unsafeTraverse(headers storage.Headers, block *flow.Header, visitor onVisitBlock, lowestHeightToVisit uint64, sanityChecker sanityCheckLowestVisitedBlock) error {
+	for {
+		err := visitor(block)
+		if err != nil {
+			return err
+		}
+
+		if block.Height == lowestHeightToVisit {
+			err = sanityChecker(block)
+			if err != nil {
+				return fmt.Errorf("lowest visited block at height %d lowestHeightToVisit failed sanity check: %w", lowestHeightToVisit, err)
+			}
+			return nil
+		}
+
+		block, err = headers.ByBlockID(block.ParentID)
+		if err != nil {
+			return fmt.Errorf("could not get block header %x: %w", block.ParentID, err)
+		}
+	}
+}
+
+// TraverseForward traverses the given fork (specified by block ID `forkHead`)
+// in the order of increasing height. The `terminal` defines when the traversal
+// stops. The `visitor` callback is called for each block in this segment.
+func TraverseForward(headers storage.Headers,
+	forkHead flow.Identifier,
+	visitor onVisitBlock,
+	terminal Terminal,
+) error {
+	lowestHeightToVisit, sanityChecker, err := terminal.Translate2Height(headers)
+	if err != nil {
+		return fmt.Errorf("error evaluating terminal: %w", err)
+	}
+
+	startBlock, err := headers.ByBlockID(forkHead)
+	if err != nil {
+		return fmt.Errorf("could not fork head with if  %x: %w", forkHead, err)
+	}
+	if startBlock.Height < lowestHeightToVisit {
+		return nil
+	}
+
+	blocks := make([]*flow.Header, 0, startBlock.Height-lowestHeightToVisit+1)
+	err = unsafeTraverse(headers,
+		startBlock,
+		func(header *flow.Header) error {
+			blocks = append(blocks, header)
+			return nil
+		},
+		lowestHeightToVisit,
+		sanityChecker,
+	)
+	if err != nil {
+		return err
+	}
+
+	for i := len(blocks) - 1; i >= 0; i-- {
+		err = visitor(blocks[i])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// functor that will be called on each block header when traversing blocks.
+type sanityCheckLowestVisitedBlock = func(header *flow.Header) error
+
+// Terminal specifies the terminal condition for traversing a fork.
+// Any condition that can be converted to a block height can be
+// represented as a terminal.
+type Terminal interface {
```

Comment:
do we actually need an interface with one function? Maybe we can get around with functor?

Reply:
We could. For some reason I can't really explain, I am trying to avoid functors in API methods. I associate this with code cleanliness ... But this is a highly-personal preference. Happy to change it too. 
In this case, I don't think it is performance relevant, as the `Terminal`'s `Translate2Height` method is only called once.

Reply:
Haha, I say that I prefer interfaces; but then I have implemented:
```golang
type Terminal interface {
	Translate2Height(headers storage.Headers) (uint64, sanityCheckLowestVisitedBlock, error)
}
```
where `sanityCheckLowestVisitedBlock` is a functor: https://github.com/onflow/flow-go/blob/9986c78f732eae64994985a4ea62d311a0f17197/state/fork/traverse.go#L104-L105

... I really don't know 🤷

Reply:
Originally I've made the comment because it's only one function and there is no sign of adding more. On other hand we will have functor that returns a tuple where one of members is another functor, so maybe the interface is actually a good choice here.